improve ledger documentation (#625)
onflow/flow-go: ledger/README.md

- **Complete Ledger** implements a fast, memory-efficient and reliable ledger. It holds a limited number of recently used states in memory (for speed) and uses write-ahead logs and checkpointing to provide reliability. Under the hood complete ledger uses a collection of MTries(forest). MTrie is a customized in-memory binary Patricia Merkle trie storing payloads at specific storage paths. The payload includes both key-value pair and storage paths are determined by the PathFinder. Forest utilizes unchanged sub-trie sharing between tries to save memory.

- **Partial Ledger** implements the ledger functionality for a limited subset of keys. Partial ledgers are designed to be constructed and verified by a collection of proofs from a complete ledger. The partial ledger uses a partial binary Merkle trie which holds intermediate hash value for the pruned branched and prevents updates to keys that were not part of proofs.

## Definitions

### binary Merkle tree

![binary Merkle tree image](/ledger/docs/binary_merkle_tree.png?raw=true "binary Merkle tree" )

> In this context a *binary Merkle tree* is defined as a full binary tree with an specific hight including three type of nodes:

Reviewer: nit: typo, s/hight/height

Author: fixed at 3e5ba56

> In this context a *binary Merkle tree* is defined as a full binary tree with an specific height including three type of nodes:

Reviewer 2: I think we should use the established term 
```suggestion
In this context a *binary Merkle tree* is defined as a [perfect binary tree](https://xlinux.nist.gov/dads/HTML/perfectBinaryTree.html) with a specific height including three type of nodes:
```

Author: updated at 0d1658b

- leaf nodes: holds a payload (data), a path (where the node is located), and a hash value (hash of path and payload content)

- empty leaf nodes: doesn't hold any data and only stores a path, and a default hash value based on the height of tree

> - intermediate nodes: holds a path and a hash value which is defined as hash of hash value of left and right children.

Reviewer 2: two comments:
* I think we should use consistent terminology and call this an `interior node` as we do in the implementation https://github.com/onflow/flow-go/blob/3e5ba56f990e1078c85c40f8aa29e37e7c12c583/ledger/complete/mtrie/node/node.go#L20-L23
* Furthermore, our interior nodes don't really hold paths.

Author: updated at 0d1658b

![node types image](/ledger/docs/node_types.png)

A *path* is a unique address of a node storing a payload. Paths are derived from the content of payloads (see common/pathfinder).

![paths image](/ledger/docs/paths.png?raw=true "paths")

#### Operations

**Get**: Fetching a payload from the binary Merkle tree is by traversing the tree based on path bits. (0: left branch, 1: right branch)

**Update**: Updates to the tree starts with traversing the tree to the leaf node, updating payload, hash value of that node and hash value of all the ancestor nodes (nodes on higher level connected to this node).

![update image](/ledger/docs/tree_update.gif?raw=true "update")

**Prove**: A binary Merkle tree can provide an inclusion proof for any given payload. A Merkle proof in this context includes all the information needed to walk through a tree branch from an specific leaf node (key) up to the root of the tree (yellow node hash values are needed for inclusion proof for the green node).

![proof image](/ledger/docs/proof.png?raw=true "proof")

### binary Merkle trie

> A **binary Merkle trie** in this context is defined as a compact version of binary Merkle tree, providing exact same functionality but don’t store unnecessary nodes.

Reviewer 2: I think it would be good to specify what "unnecessary nodes" are:

```suggestion
A **binary Merkle trie** in this context is defined as a compact version of binary Merkle tree, providing exact same functionality but doesn't explicitly store empty nodes. Formally, a node is empty: 
* the node is an empty leaf node (see definition above)
* an interior node is defined to be empty, if its two children are empty. 
```

Author: updated at 0d1658b

![binary partial trie image](/ledger/docs/trie_update.gif?raw=true "binary partial trie")

### forest 

> A **forest** holds a set of binary Merkle tries, given a parent trie and a batch of updates it creates a new binary merkle trie and adds it to the forest.

Reviewer 2: This is how we _use_ the forrest. However, a forest is just a specific type of graph. There is no concept of updates contained in the graph-theoretical definition:
> [[source](https://mathworld.wolfram.com/Forest.html)] A forest is an acyclic graph. Forests therefore consist only of (possibly disconnected) trees, hence the name "forest." 

```suggestion
Formally, a **forest** is any acyclic graph. Any set of disjoint trees forms a forest. In the context of Flow, we take an existing state, represented by a Merkle tree. Updating the payload of some of the leafs creates a new Merkle tree, which we add to the forest. In other words, the forest holds a set of state snapshots
```

Author: updated at 0d1658b

> ![forest image](/ledger/docs/forest.png?raw=true "forest")

Reviewer 2: I find this figure potentially misleading. It looks like there are edges connecting the different forests (same solid line as used for edges within the trees). But here, the lines don't represent graph edges, but "update relations". I would suggest to use a different line style (like dotted lines).

Author: I removed the image it was not really expressive.

- **Complete Ledger** implements a fast, memory-efficient and reliable ledger. It holds a limited number of recently used states in memory (for speed) and uses write-ahead logs and checkpointing to provide reliability. Under the hood complete ledger uses a collection of MTries(forest). MTrie is a customized in-memory binary Patricia Merkle trie storing payloads at specific storage paths. The payload includes both key-value pair and storage paths are determined by the PathFinder. Forest utilizes unchanged sub-trie sharing between tries to save memory.

- **Partial Ledger** implements the ledger functionality for a limited subset of keys. Partial ledgers are designed to be constructed and verified by a collection of proofs from a complete ledger. The partial ledger uses a partial binary Merkle trie which holds intermediate hash value for the pruned branched and prevents updates to keys that were not part of proofs.

## Definitions
In this section we provide an overview of some of the concepts. Hence it is highly recommended to checkout [this doc](https://github.com/onflow/flow-go/blob/master/ledger/complete/mtrie/README.md) for the formal and technical definitions in more details.

> ### binary Merkle tree

Reviewer 3: I suggest to use this terminology to be consistent with we define [here](https://github.com/onflow/flow-go/blob/master/ledger/complete/mtrie/README.md#storage-model):
- Merkle trie is the binary radix tree we would like to model. 
- Mtrie (memory-trie) is our implementation of a Merkle trie, with all the optimizations and shortcuts.

```suggestion
### binary Merkle trie
```

![binary Merkle tree image](/ledger/docs/binary_merkle_tree.png?raw=true "binary Merkle tree" )

> In this context a *binary Merkle tree* is defined as [perfect binary tree](https://xlinux.nist.gov/dads/HTML/perfectBinaryTree.html) with a specific height including three type of nodes:

Reviewer 3: Classifying three types of nodes is not consistent with what is defined [here](https://github.com/onflow/flow-go/blob/master/ledger/complete/mtrie/README.md#underling-graph-theoretical-storage-model) (we explicitly say there are 2 types of nodes). I suggest to remove `empty leaf nodes` as we define them as `leaf nodes` for a Merkle trie. 

We only consider empty leafs when we represent a "Merkle trie" in memory as a "memory-trie" (Mtrie).

Author: removed at 2ba9d6c

- leaf nodes: holds a payload (data), a path (where the node is located), and a hash value (hash of path and payload content)

- empty leaf nodes: doesn't hold any data and only stores a path, and a default hash value based on the height of tree

- interim nodes: holds a hash value which is defined as hash of hash value of left and right children.

![node types image](/ledger/docs/node_types.png)

> A *path* is a unique address of a node storing a payload. Paths are derived from the content of payloads (see common/pathfinder).

Reviewer 3: I suggest to clarify what a path is in our case.

```suggestion
A *path* is a unique address of a node storing a payload. Paths are derived from the content of payloads (see common/pathfinder).
A path is explicitly a hash of 256 bits.  
```

Author: Applied at 2ba9d6c

![paths image](/ledger/docs/paths.png?raw=true "paths")

#### Operations

**Get**: Fetching a payload from the binary Merkle tree is by traversing the tree based on path bits. (0: left branch, 1: right branch)

**Update**: Updates to the tree starts with traversing the tree to the leaf node, updating payload, hash value of that node and hash value of all the ancestor nodes (nodes on higher level connected to this node).

![update image](/ledger/docs/tree_update.gif?raw=true "update")

**Prove**: A binary Merkle tree can provide an inclusion proof for any given payload. A Merkle proof in this context includes all the information needed to walk through a tree branch from an specific leaf node (key) up to the root of the tree (yellow node hash values are needed for inclusion proof for the green node).

![proof image](/ledger/docs/proof.png?raw=true "proof")

> ### binary Merkle trie

Reviewer 3: Here comes how we model the Merkle trie:

```suggestion
### Memory-trie (Mtrie)
```

Author: Applied at 2ba9d6c

> A **binary Merkle trie** in this context is defined as a compact version of binary Merkle tree, providing exact same functionality but doesn't explicitly store empty nodes. Formally, a node is empty: 

Reviewer 3: ```suggestion
An **Mtrie** in this context is defined as a compact version of binary Merkle trie, providing exact same functionality but doesn't explicitly store empty nodes. Formally, a node is empty: 
```

Author: Applied at 2ba9d6c

> * the node is an empty leaf node (see definition above)

Reviewer 3: Once we remove the empty leaf node from above, we would need to define it here:

```suggestion
* the node is an empty leaf node: it doesn't hold any data and only stores a path. Its hash value is defined as a default hash based on the height of tree.
```

Author: Applied at 2ba9d6c

* an interim node is defined to be empty, if its two children are empty. 

![binary partial trie image](/ledger/docs/trie_update.gif?raw=true "binary partial trie")

### forest 
Formally, a **forest** is any acyclic graph. Any set of disjoint trees forms a forest. In the context of Flow, we take an existing state, represented by a Merkle tree. Updating the payload of some of the leafs creates a new Merkle tree, which we add to the forest. In other words, the forest holds a set of state snapshots

### compact forest 
A **compact forest** constructs a new trie after each update (copy on change) and reuses unchanged sub-tries from the parent.

![compact forest image](/ledger/docs/reuse_sub_trees.gif?raw=true "compact forest")

### path finder 

> **Path finder** deterministically computes a path for a given payload. Path finder is responsible to make sure the trie grows in balance and relevant data are stored nearby so that the proof size can be optimized.

Reviewer 3: I think the "trie grows in balance" point and "data stored nearby" go in opposite directions : if we prioritize a balanced tree then we loose storage locality, and if we prioritize data locality we will end up with an unbalanced tree.
 
I suggest we remove the storage locality point since the implemented version of path finder is just guaranteeing a balanced tree through hashing (at least for now).  

```suggestion
**Path finder** deterministically computes a path for a given payload. Path finder is responsible for making sure the trie grows in balance.
```

Author: Applied at 2ba9d6c
